lib/archethic_web/controllers/api/transaction_controller.ex

```diff
@@ -143,4 +143,80 @@ defmodule ArchethicWeb.API.TransactionController do
         |> render("400.json", changeset: changeset)
     end
   end
+
+  @doc """
+  This controller, Fetch the recipients contract and simulate the transaction, managing possible
+  exits from contract execution
+  """
+  def simulate_contract_execution(
+        conn,
+        params = %{}
+      ) do
+    case TransactionPayload.changeset(params) do
+      changeset = %{valid?: true} ->
+        tx =
+          %Transaction{data: %TransactionData{recipients: recipients}} =
+          changeset
+          |> TransactionPayload.to_map()
+          |> Transaction.cast()
+
+        {:ok, sup} = Task.Supervisor.start_link(strategy: :one_for_one)
+
+        results =
+          Task.Supervisor.async_stream_nolink(
+            sup,
+            recipients,
+            fn recipient ->
+              fetch_recipient_tx_and_simulate(recipient, tx)
+            end,
+            on_timeout: :kill_task
+          )
+          |> Stream.map(fn
+            {:ok, result} ->
+              result
+
+            nil ->
+              %{"valid" => false, reason: "A contract timed out"}
+
+            {:exit, reason} ->
+              Logger.debug(
+                "Contract #{inspect(tx)} exited while simulating  : #{inspect(reason)}"
+              )
+
+              %{"valid" => false, reason: "A contract exited while simulating"}
+          end)
+          |> Enum.to_list()
+
+        conn
+        |> put_status(:ok)
+        |> json(results)
+
+      changeset ->
+        error_details =
+          Ecto.Changeset.traverse_errors(changeset, &ArchethicWeb.ErrorHelpers.translate_error/1)
+
+        json_body =
+          Map.merge(error_details, %{"valid" => false, reason: "Query validation failled"})
+
+        conn
+        |> put_status(:ok)
+        |> json([json_body])
+    end
+  end
+
+  defp fetch_recipient_tx_and_simulate(recipient, tx) do
+    case Archethic.search_transaction(recipient) do
+      {:ok, prev_tx} ->
+        case Archethic.simulate_contract_execution!(prev_tx, tx) do
+          :ok ->
+            %{:valid => true, :address => Base.encode64(tx.address)}
+
+          {:error, reason} ->
+            %{"valid" => false, :address => Base.encode64(tx.address), reason: reason}
+        end
+
+      {:error, reason} ->
+        %{"valid" => false, :address => Base.encode64(tx.address), reason: reason}
```

Comment:
We usually use hexadecimal to represent `address` in a json (encode16).
Use only strings for the keys in the `=>` map notation

I did not manage to test it via the endpoint. All I get is `invalid_inherit_constraints`.
I am using this contract: 
```
condition inherit: [
	content: true
]

condition transaction: [
	uco_transfers: size() > 0
]

actions triggered_by: transaction do
	set_content "hello"
end
```

Reply:
I think I found the issue. It's because you send the "incoming transaction" in `simulate_contract_execution`. But `simulate_contract_execution` wants the "next transaction" (you get that one after executing the contract (or nil)).